firefox-ios/Client/Frontend/Browser/SearchViewController.swift

```diff
@@ -847,6 +838,76 @@ class SearchViewController: SiteTableViewController,
                                  indexPath)
     }
 
+    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+        if let section = SearchListSection(rawValue: indexPath.section) {
```

Comment:
I'm a little nervous about using `tableView(_:willDisplay:forRowAt:)`, and was wondering if you've seen any issues with it.

When I was working on Firefox Suggestion impressions for visible suggestions in [this PR](https://github.com/mozilla-mobile/firefox-ios/pull/18273), I noticed that indexing into `self.data`, `self.suggestions`, `self.filteredOpenTabs`, `self.firefoxSuggestions`, etc. from `tableView(_:willDisplay:forRowAt:)` would intermittently crash the app with an index out-of-bounds exception. Typing quickly into the address bar, then deleting and changing my input, was enough to crash it.

I wrote a note [here](https://github.com/mozilla-mobile/firefox-ios/pull/18409#issue-2101448322), but I wonder if the crash is caused by a race between the callbacks that update `self.filteredOpenTabs`, `self.data`, etc. and `tableView.reloadData()`. By the time `tableView(_:willDisplay:forRowAt:)` is called, the data might have been updated again, but the delegate method is still invoked with the old value, and so the `indexPath` is no longer valid.

For `search_suggestion_impressed`, I fixed that by only reading `tableView.visibleIndexPaths` (via `SearchViewController.visibleSuggestionsTelemetryInfo`) in `SearchViewControllerDelegate.searchViewControllerWillHide`. I think we could still end up with a race here (#18409 again), but the index paths should be settled.

I'm not sure if that'll work quite the same here for `SearchTelemetry`, though... 🤔

Reply:
Update: I didn't see any crashes when interacting with the awesomebar in testing, so I was nervous for no good reason ☺️ I must have been doing something else wrong when I was working on #18273, sorry for the noise!